src/pytest_datadir/plugin.py

```diff
@@ -55,3 +57,76 @@ def datadir(original_datadir, tmp_path):
     else:
         result.mkdir()
     return result
+
+
+@dataclass(frozen=True)
+class LazyDataDir:
+    """
+    A dataclass to represent a lazy data directory.
+    It is used to ensure that the directory is created only when needed.
+    """
+
+    original_datadir: Path
+    tmp_path: Path
+
+    def joinpath(self, other: str) -> Path:
+        """Helper method to join paths with lazy loading.
+
+        Args:
+            other (str): The path to join with the lazy data directory.
+
+        Returns:
+            Path: The joined path, without repeat copying any files.
+        """
+        original = self.original_datadir / other
+        target = self.tmp_path / other
+        if not target.exists():
```

Comment:
It is OK to use `lazy_datadir.joinpath("new file")` when `new file` does not exist in the data directory:

```suggestion
        if original.exists() and not target.exists():
```

Reply:
I think that the `is_file` and `is_dir` calls just below are already requiring that `original` already exists? That is, they return false even if the specified path _would_ point to a file/directory, but that file/directory doesn't exist. So I don't see what this suggested code change would actually change.

Reply:
Oh you are right of course.

But on second look I think that check there makes the logic more explicit.

But up to you to apply it or leave it out. :+1: